### ISSUE TITLE
High duration loss

### ISSUE THREAD

**Issue by w11wo** [p0p4k/pflowtts_pytorch#40]
Hi @p0p4k, thanks for making this repo!

I am currently trying to train a 44.1kHz English model, but my model is struggling with a rather high duration loss when compared against your TensorBoard logs. It currently looks as follows:

![image](https://github.com/p0p4k/pflowtts_pytorch/assets/23167175/38684e71-deb4-4aa1-ad38-d69ef9e41791)

It seems like the other loss terms are correct.

Also, when the generated mel-spectrogram is passed to a vocoder, the audio is very much wrong in pronunciation -- maybe only half right.

My P-Flow config can be found [here](https://github.com/bookbot-hive/pflowtts_pytorch/blob/eng/configs/data/en_au_dean2zak.yaml), and the corresponding HiFi-GAN vocoder config can be found [here](https://github.com/bookbot-hive/hifi-gan/blob/master/config_v4.json).

Could you please let me know where I might be wrong? Thanks in advance!

**Comment by Oleksandr2505** [comment]
> Hi @p0p4k, thanks for making this repo!
> 
> I am currently trying to train a 44.1kHz English model, but my model is struggling with a rather high duration loss when compared against your TensorBoard logs. It currently looks as follows:
> 
> ![image](https://private-user-images.githubusercontent.com/23167175/316711611-38684e71-deb4-4aa1-ad38-d69ef9e41791.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTE0MjYxNzYsIm5iZiI6MTcxMTQyNTg3NiwicGF0aCI6Ii8yMzE2NzE3NS8zMTY3MTE2MTEtMzg2ODRlNzEtZGViNC00YWExLWFkMzgtZDY5ZWY5ZTQxNzkxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMjYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzI2VDA0MDQzNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPThmNDk1ZTdmZDY1MGQzMmVjMWFlN2NhYjk0ZGY2Mzk0NzVjNWU0NDE0ODU4MzE2ZjQ2OWU3YzNlNjI3N2YwZjUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0._Ld8TM5AM8eb-XXWTiE8UI8vgGrX4hq3b-xV-ssoHCg)
> 
> It seems like the other loss terms are correct.
> 
> Also, when the generated mel-spectrogram is passed to a vocoder, the audio is very much wrong in pronunciation -- maybe only half right.
> 
> My P-Flow config can be found [here](https://github.com/bookbot-hive/pflowtts_pytorch/blob/eng/configs/data/en_au_dean2zak.yaml), and the corresponding HiFi-GAN vocoder config can be found [here](https://github.com/bookbot-hive/hifi-gan/blob/master/config_v4.json).
> 
> Could you please let me know where I might be wrong? Thanks in advance!

Hello, Make sure u wrote down "english_cleaners2" in synthesise.py 
it has to be like this when u generate english speech

sequence = torch.tensor(intersperse(text_to_sequence(stressed_text, ['english_cleaners2']), 0), dtype=torch.long,

**Comment by w11wo** [comment]
Hi @Oleksandr2505. 

Yes I have changed both the training and inference phonemizer to `english_cleaners3`, so it's definitely not that issue. I've also checked the phonemization output and it is correct.

It's also unrelated to the issue I raised, which is more on the training duration loss, not inference.

**Comment by Oleksandr2505** [comment]
> Hi @Oleksandr2505.
> 
> Yes I have changed both the training and inference phonemizer to `english_cleaners3`, so it's definitely not that issue. I've also checked the phonemization output and it is correct.
> 
> It's also unrelated to the issue I raised, which is more on the training duration loss, not inference.

I am training my model on english 4h 22050kHz dataset, I've recently passed 1000epochs and it sounds good. Maybe you should try switch the vocoder, because at the start I had a huge artifacts, I took vocoder trained on VCTK v1 dataset and it became better. Maybe you should look for another vocoder to match 44 sample rate. You also texted about training duration loss, is it something different from regular loss?
Btw this is my logs.
 [sample](![image_2024-03-26_061754446](https://github.com/p0p4k/pflowtts_pytorch/assets/70656397/2fa9a68c-a5d7-4e5f-999a-020b25673f5c))

**Comment by w11wo** [comment]
Yeah, your loss curves look very much like mine.

I'd say I probably don't have an issue with the vocoder, since it's not an issue of artifacts, but pronunciation quality. And yes, the HiFi-GAN I'm using has been trained on 44.1kHz, which isn't a problem.

My main suspect is from my rather high duration losses (~2), versus the TensorBoard graphs posted on the README which can easily be < 1.0 early on, which I thought was the expected trend. It's likely that this affects the pronunciation quality too.

I just wanted to make sure that my 44.1kHz P-Flow config is correct, in case I'm missing something. But thanks for your suggestions nonetheless.

**Comment by Oleksandr2505** [comment]
Yea, Actually I see that author got less then 1, but we don't know his exact config. So we can only test and try, I have this parameter very close 1 as well, eventually it can get under it, but can't know on 100%. One more tip, when I was training ukrainian languages and then tried to run english models, I was getting half right and wrong words It was mainly cause of cleaners, but I was changing something in pflow/text in folder like "symbols" and other. So make sure you did not add something another instead of english symbols or numbers. I've fixed in this way.

![image_2024-03-26_063436405](https://github.com/p0p4k/pflowtts_pytorch/assets/70656397/a92f4e2e-39c6-491c-8ad0-37d70871630c)

**Comment by Oleksandr2505** [comment]
> Yeah, your loss curves look very much like mine.
> 
> I'd say I probably don't have an issue with the vocoder, since it's not an issue of artifacts, but pronunciation quality. And yes, the HiFi-GAN I'm using has been trained on 44.1kHz, which isn't a problem.
> 
> My main suspect is from my rather high duration losses (~2), versus the TensorBoard graphs posted on the README which can easily be < 1.0 early on, which I thought was the expected trend. It's likely that this affects the pronunciation quality too.
> 
> I just wanted to make sure that my 44.1kHz P-Flow config is correct, in case I'm missing something. But thanks for your suggestions nonetheless.

Yea, Actually I see that author got less then 1, but we don't know his exact config. So we can only test and try, I have this parameter very close 1 as well, eventually it can get under it, but can't know on 100%. One more tip, when I was training ukrainian languages and then tried to run english models, I was getting half right and wrong words It was mainly cause of cleaners, but I was changing something in pflow/text in folder like "symbols" and other. So make sure you did not add something another instead of english symbols or numbers. I've fixed in this way.
![image_2024-03-26_064010400](https://github.com/p0p4k/pflowtts_pytorch/assets/70656397/001dc0bb-b710-4bf5-b8d5-591041df3af9)

**Comment by Oleksandr2505** [comment]
> Yeah, your loss curves look very much like mine.
> 
> I'd say I probably don't have an issue with the vocoder, since it's not an issue of artifacts, but pronunciation quality. And yes, the HiFi-GAN I'm using has been trained on 44.1kHz, which isn't a problem.
> 
> My main suspect is from my rather high duration losses (~2), versus the TensorBoard graphs posted on the README which can easily be < 1.0 early on, which I thought was the expected trend. It's likely that this affects the pronunciation quality too.
> 
> I just wanted to make sure that my 44.1kHz P-Flow config is correct, in case I'm missing something. But thanks for your suggestions nonetheless.

Btw, I would like to know, does your models synthesis the descent sufficient pauses between the sentences or not? I could not find how to control it

**Comment by w11wo** [comment]
@Oleksandr2505 I can't tell, to be honest. The pronunciation isn't good enough to determine the word pauses. But per my experiments with other TTS models (VITS, VITS2), you just have to train with audio that has the long enough pauses that you want.

If you want models with more customizable pauses, I'd recommend something like FastSpeech2 which is more deterministic since it does a more deterministic duration prediction, I think. VITS/VITS2/P-Flow is better for more expressive speech.

**Comment by Oleksandr2505** [comment]
@w11wo 

thank you for info!

**Comment by w11wo** [comment]
Update: I tried changing many different setups, but found quite an interesting change that *finally* led to a decent performance.

The duration loss (which contributes to the overall loss), converges to a lower value faster if the model's vocab size is 178, derived from the default LJSpeech + espeak setup. In the logs above where I used gruut as the phonemizer, I used a smaller vocab size of 78 in the model config. Simply increasing the vocab size back to 178 led to a better convergence of the duration loss (for some reason). However, I'm still unable to get a loss of < 1 unlike the one posted in the README.

**Comment by w11wo** [comment]
@patriotyk Yes, I do hear a massive difference in audio output. I think the duration loss values reflect the overall performance of the audio as well, given that the validation loss also decreases with the increased vocab size.

Eventually I could get the duration loss to ~1.0, which is much better than the initial experiments with the smaller vocab size.

The odd thing is *why* vocab size even impacts the duration loss -- I'm clueless about this.

**Comment by patriotyk** [comment]
I have found that `configs/model/pflow.yaml` contains `n_vocab` that is equal to 178. I thing it should be 78 in your case and maybe you will get even better duration loss because all 78 symbols are real and you will not have unused symbols.

**Comment by w11wo** [comment]
@patriotyk not really. What I meant by my comments above is that I started with vocab size of 78, which led to the high loss graph as attached at the top of this issue post. It is only by increasing it to 178 that I could achieve a low duration loss of about 1.

**Comment by w11wo** [comment]
Ah yeah. Whenever I changed the symbol list, I also made sure to change the vocab size value in the config. It's weird how an actual symbol size of 78 would benefit by increasing the model config's vocab size.

**Comment by Oleksandr2505** [comment]
@w11wo Hi, i just recalled u mentioned about training model using hifigan 44.1khz vocoder, where did you get it or how train it?

**Comment by w11wo** [comment]
Hi @Oleksandr2505. I ended up training a 44.1kHz HiFi-GAN from scratch, which took about 6 days on a 1xH100 GPU. You can find our fork of the HiFi-GAN training code/repo [here](https://github.com/bookbot-hive/hifi-gan).
